drivers/serial/uart_stm32.c

```diff
@@ -1628,11 +1614,15 @@ static int uart_stm32_init(const struct device *dev)
 		}
 	}
 
-	/* Configure dt provided device signals when available */
-	err = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
-	if (err < 0) {
-		return err;
-	}
+	return 0;
+}
+
+static int uart_stm32_registers_configure(const struct device *dev)
```

Comment:
Currently `uart_stm32_configure` doesn't save any application configuration in instance data.
Hence, if application performed a runtime configuration before a power state where registers are reset, this configuration will be lost so need to be performed again.
Do you think this can be performed by application ? (which implies that application needs to keep track of device state and needs)
If runtime configuration restoration has to be performed by device itself, it needs to be able to store it and re-apply when needed.

Applicability of this point depends on runtime configuration potential of the device API in use.

Reply:
That's a very good point, I hadn't considered this.
My use case for the UART changes was to ensure that the serial console output worked when the SoC comes out of STOP2, in order to assist debugging. :smile:
AFAICT, general usage of `CONFIG_UART_CONSOLE` doesn't make use of the `uart_configure` syscall, which would trigger `uart_stm32_configure`, but please correct me if I'm wrong.

I don't think the application should keep track of, or be required to restore the UART runtime configuration, that feels more like the scope of the OS services.

My first go at this would be to add a struct to hold the runtime configuration, and if that's been initialized by an application calling `uart_configure` at some point, the UART reinit function would prefer restoring the parameters stored there rather than the generic register initialization also done on boot.

There's significant overlap between the newly added `uart_stm32_registers_configure` functionality, originally taken from `uart_stm32_init`, and `uart_stm32_configure`, so I'll try to unify the behavior of these with the new struct in a way that's more effective and less redudant than the current code.

Reply:
> My first go at this would be to add a struct to hold the runtime configuration, and if that's been initialized by an application calling uart_configure at some point, the UART reinit function would prefer restoring the parameters stored there rather than the generic register initialization also done on boot.

Sounds reaonable

> There's significant overlap between the newly added uart_stm32_registers_configure functionality, originally taken from uart_stm32_init, and uart_stm32_configure, so I'll try to unify the behavior of these with the new struct in a way that's more effective and less redudant than the current code.

Good point. Don't hesitate to split this work in a different PR to focus current PR on its initial topic.

Reply:
> Good point. Don't hesitate to split this work in a different PR to focus current PR on its initial topic.

Since the changes are tightly coupled with the reinit logic used here, I've decided to amend the commit with my above suggestion. If you think it muddies the scope of this PR too much, and you'd prefer I put it in a separate PR, just let me know and I'll oblige.

The commit message describes the changes pretty well I think, but here's some key points to aid review:

- Added static `uart_config` struct to `STM32_UART_INIT` macro
- Added `uart_config` struct pointer as `uart_stm32_config` member
- (Re)moved now superfluous `baud_rate`, `parity`, and `hw_flow_control` struct members
- Moved overlapping baudrate/parity/data-/stop-bits/flow control configuration to common function, with persistence in static instance `uart_config` struct
- Removed incorrect runtime parity/data bits configuration, and replaced with thoroughly tested compile-time `BUILD_ASSERT` statements
  - Original runtime logic in `init` forced a very small subset of UART configurations, and straight-up ignored many valid DT property combinations (e.g 7E1)
  - This now also allows invalid configurations to get flagged compile-time
  - The previous logic partially conflicted with the logic in `uart_stm32_configure`, especially that relating to parity/data-bits combinations, either enforced opposite configurations
- The `#if defined(LL_USART_STOPBITS_1_5) && HAS_LPUART_1` conditional broke "half-bit" parity application configurations of USARTs, this is now fixed

Overall the driver will now behave more coherently, regardless if configured by DTS on (re-)init, or by application/subsystem syscall.

Reply:
Some follow up questions on things that don't make sense to me, and might need attention as well:

- Is there any reason the driver forbids runtime configuration with [9 data bits](https://github.com/zephyrproject-rtos/zephyr/blob/89dabaeec2c2485c0a256ee6465dd5c0878a4fc8/drivers/serial/uart_stm32.c#L523-L524)? A quick grep of the HAL module shows all STM32 series UART support 9 data bits.
- Why does the runtime configuration [default to 2 stop bits](https://github.com/zephyrproject-rtos/zephyr/blob/89dabaeec2c2485c0a256ee6465dd5c0878a4fc8/drivers/serial/uart_stm32.c#L320-L321) instead of 1?

Reply:
> Some follow up questions on things that don't make sense to me, and might need attention as well:

> Is there any reason the driver forbids runtime configuration with [9 data bits](https://github.com/zephyrproject-rtos/zephyr/blob/89dabaeec2c2485c0a256ee6465dd5c0878a4fc8/drivers/serial/uart_stm32.c#L523-L524)? A quick grep of the HAL module shows all STM32 series UART support 9 data bits.
Why does the runtime configuration [default to 2 stop bits](https://github.com/zephyrproject-rtos/zephyr/blob/89dabaeec2c2485c0a256ee6465dd5c0878a4fc8/drivers/serial/uart_stm32.c#L320-L321) instead of 1?

This might actually be a genuine bug. See related https://github.com/zephyrproject-rtos/zephyr/issues/56003 for instance

Reply:
There's certainly a good effort to document the change on this commit, but I'd prefer this is split in a dedicated PR (and potentially few commits if doable).
This will definitely ease and speed up the review (avoid mixing topics, ..).
Please also note related change https://github.com/zephyrproject-rtos/zephyr/pull/54354 that might get across your changes

Reply:
Sure thing, I'll move all UART changes into a separate PR when I find a spare few minutes today. 

That means this PR won't touch the UART driver until it's merged. If this should get merged first, I'll have to add the UART reinit calling via `pm_action` separately–but that should be very similar to the other drivers', based on my proposed implementation of this mentioned below.